### DISCUSSION TITLE
Solving transmission line with "Circuit"

### DISCUSSION THREAD

**Discussion by krono-i2** [scikit-rf/scikit-rf#646] [Q&A] [answered]
Hi guys!
First, I want to thank all of contributors; scikit-rf is great for both research and academic tutorial purposes!
I've some questions, but I want also show what we found out about `Circuit` behiviour for solving transmission lines.

Let's suppose I want to solve the simple transmission line you stated [here](https://scikit-rf.readthedocs.io/en/latest/examples/circuit/Voltages%20and%20Currents%20in%20Circuits.html), considering a fully non-matching scenario (i.e., `Z_S!=Z_0`, and `Z_0!=Z_L`), with respect to the voltage provided by the generator (defined as `V_S`).

```
Z_0 = 50+20j
Z_S = 50+30j
Z_L = 50+10j
L = 10
V_S = 1+1j

freq = rf.Frequency(2,2,1,unit="GHz")
line_media = rf.media.DefinedGammaZ0(freq,Z0=Z_0)
line = line_media.line(d=L, unit="m", name="line")
```

I can simply use tlineFunctions to manually solve my problem carrying out the load transport, the voltage divider, finally, the voltage/current propagation to the load.

```
theta = rf.theta(line_media.gamma, freq.f, L)
Z_in = rf.zl_2_zin(Z_0, Z_L, theta=theta)
V_in = V_S*Z_in/(Z_in+Z_S)
I_in = V_S/(Z_in+Z_S)

print('Input voltage and current:',V_in,'V and',I_in,'A')

V_out, I_out = rf.tlineFunctions.voltage_current_propagation(V_in, I_in, Z_0, theta)
print('Output voltage and current:',V_out,'V and',I_out,'A')
```

For more complex trasmission lines it could be useful resort `Circuit`.
Now the questions:
Q1) What's the "port impedance for media" `z0` in `skrf.media.DefinedGammaZ0()`?
Q2) What's the "characteristic impedence of the port" `z0` in `skrf.circuit.Port()`? Why does it need to be real?
At the moment I'm ignoring Q1, and we found out that if I set `z0 = Z_S` (complex!) for the excited port (port1) and `z0 = Z_L` (complex too!) for the load port (port2), I obtain the same results of tlineFunctions when I use a "strange" formula for calculating `P_f` from `V_S` and `Z_S`.

```
# Using Circuit
P_f = 1/(8*Z_S)*V_S**2
print("The Forward Power (complex?) will be: ",P_f)
port1 = rf.Circuit.Port(frequency=freq, name="port1", z0=Z_S)
port2 = rf.Circuit.Port(frequency=freq, name="port2", z0=Z_L)
cnx = [
    [(port1, 0), (line, 0)],
    [(port2, 0), (line, 1)]
]
crt = rf.Circuit(cnx)

power = [P_f, 0]
phase = [0, 0]

V_at_ports = crt.voltages_external(power, phase)
print(V_at_ports)

I_at_ports = crt.currents_external(power, phase)
print(I_at_ports)
```

What's the story here? I'm supposing I'm exploiting a bug to obtain the exactly result I want!
Thanks for help in advance!

Ivan

**Comment by jhillairet** [comment, marked as answer]
Dear Ivan,

Thanks for your words.

Q1) The port impedance `z0` in `Media` is used to compute an impedance mismatch if any Networks of different port impedance are connected. 
Setting the port impedance `z0` is only needed if it should be different from the characteristic impedance `Z0` of the media. (If `z0` is `None` then it will be default to `Z0`.). Example:
```python
>>>rf.DefinedGammaZ0(rf.Frequency(1,1,1), Z0=70).line(1)
# 2-Port Network: '',  1.0-1.0 GHz, 1 pts, z0=[70.+0.j 70.+0.j]

>>> rf.DefinedGammaZ0(rf.Frequency(1,1,1), Z0=70, z0=2).line(1)
# 2-Port Network: '',  1.0-1.0 GHz, 1 pts, z0=[2.+0.j 2.+0.j]
```

Q2) Well, _a priori_, nothing forbid to use complex values for the characteristic impedances of the `Circuit.Port`:  behind the hood, `Port` creates a perfect match for a media of the given characteristic impedance. For such a simple case it should works. However, for more complex cases, I would not swear it works.

**Reply by jhillairet** [reply]
@krono86 , see also #651. I admit the present API can be confusing for many, me included. The API is prone to change in the future release and comments/suggestions are always welcome.
